lib/srv/desktop/rdp/rdpclient/src/piv.rs

```diff
@@ -0,0 +1,417 @@
+        //
+        // P1=04 and P2=00 means selection of DF (usually) application by name. Everything else not
+        // supported.
+        if cmd.p1 != 0x04 && cmd.p2 != 0x00 {
+            return Ok(Response::new(Status::NotFound));
+        }
+        if !PIV_AID.matches(cmd.data()) {
+            return Ok(Response::new(Status::NotFound));
+        }
+
+        // See https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf section
+        // 3.1.1
+        let resp = tlv(
+            TLV_TAG_PIV_APPLICATION_PROPERTY_TEMPLATE,
+            Value::Constructed(vec![
+                tlv(
+                    TLV_TAG_AID,
+                    Value::Primitive(vec![0x00, 0x00, 0x10, 0x00, 0x01, 0x00]),
+                )?,
+                tlv(
+                    TLV_TAG_COEXISTENT_TAG_ALLOCATION_AUTHORITY,
+                    Value::Constructed(vec![tlv(
+                        TLV_TAG_AID,
+                        Value::Primitive(PIV_AID.truncated().to_vec()),
+                    )?]),
+                )?,
+            ]),
+        )?;
+        Ok(Response::with_data(Status::Success, resp.to_vec().into()))
+    }
+
+    fn handle_verify(&mut self, _cmd: Command<S>) -> RdpResult<Response> {
+        // No PIN verification needed.
+        Ok(Response::new(Status::Success))
+    }
+
+    fn handle_get_data(&mut self, cmd: Command<S>) -> RdpResult<Response> {
+        // See https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf section
+        // 3.1.2.
+        if cmd.p1 != 0x3F && cmd.p2 != 0xFF {
+            return Ok(Response::new(Status::NotFound));
+        }
+        let request_tlv = Tlv::from_bytes(cmd.data())
+            .or_else(|e| Err(invalid_data_error(&format!("TLV invalid: {:?}", e))))?;
+        if *request_tlv.tag() != tlv_tag(0x5C)? {
+            return Ok(Response::new(Status::NotFound));
+        }
+        match request_tlv.value() {
+            Value::Primitive(tag) => match to_hex(&tag).as_str() {
+                // Card Holder Unique Identifier.
+                "5FC102" => Ok(Response::with_data(Status::Success, self.chuid.clone())),
+                // X.509 Certificate for PIV Authentication
+                "5FC105" => {
+                    self.pending_response = Some(Cursor::new(self.piv_auth_cert.clone()));
+                    self.handle_get_response(cmd)
+                }
+                _ => {
+                    // Some other unimplemented data object.
+                    Ok(Response::new(Status::NotFound))
+                }
+            },
+            Value::Constructed(_) => Ok(Response::new(Status::NotFound)),
+        }
+    }
+
+    fn handle_get_response(&mut self, _cmd: Command<S>) -> RdpResult<Response> {
+        match &mut self.pending_response {
+            None => Ok(Response::new(Status::NotFound)),
+            Some(cursor) => {
+                let mut chunk = [0; 256];
```

Comment:
Why is this 256 precisely? comment? might also be worth defining a const at the start of the function with a descriptive name since it's used a couple of times